src/Entity/OgMembership.php

```diff
@@ -83,7 +84,19 @@ class OgMembership extends ContentEntityBase implements OgMembershipInterface {
    * {@inheritdoc}
    */
   public function getCreatedTime(): int {
-    return $this->get('created')->value;
+    // Only use $this->get() if it is already populated. If it is not available
+    // then use the raw value. This field is not translatable so we do not need
+    // the slow field definition lookup from $this->getTranslatedField().
+    if (isset($this->fields['created'][LanguageInterface::LANGCODE_DEFAULT])) {
```

Comment:
I'd use a private helper method to avoid repeating the if/else.

Reply:
I considered this at some point but I ended up not doing it because there are many subtle differences that make it less reusable. It looks on the surface like repeated code, but there are small unique variations in most of them.

For example some fields have an integer as the default value (e.g. `::getCreatedTime()`), others have a string (`::getGroupEntityType()`) or an array (`::getRolesIds()`) or `NULL` because they are optional (`::getOwnerId()`).

And also the way that the property is accessed is often subtly different, for example it can be a multivalue array which is retrieved using `->getValue()` (e.g. `::getRolesIds()`), the `->value` property (`::getGroupBundle()`) or the `->target_id` property (`::getOwnerId()`).

I will give this an attempt. I can maybe capture these variations in two or three parameters and skip the strict type checking.

Reply:
Indeed, I noticed some of the differences and we could probably solve them with argumentsas you mentioned. That said, I never really liked the ambiguity within core's entity "API" (get vs magic properties vs public API) and the different API's for fields. To me this feels like we're missing some generic API to get a field value.

If there is a performance hit to using get() vs using the raw value, shouldn't this be solved higher up (in core)?

Are we 100% sure these entities will never be translated (and if so shouldn't we prevent it from being marked as translatable - if this isn't the case yet)?

Reply:
> Indeed, I noticed some of the differences and we could probably solve them with arguments as you mentioned. That said, I never really liked the ambiguity within core's entity "API" (get vs magic properties vs public API) and the different API's for fields. To me this feels like we're missing some generic API to get a field value.

Yes it definitely needs improvements, and it is currently very confusing to use, and also slow. I think the best generic API that we have is `::get()->getValue()`. It will reliably return the field values, but in order to do this it loads *ALL* the field and field storage definitions (even if you only access one single field). Then it inspect them to determine whether the field is translatable, and then the `FieldItemList` object is constructed and cached on the entity to speed up subsequent lookups.

It is understandable that core needs to do all this in its base class (`ContentEntityBase`) even though it is slow, since the field definitions are dynamic. It would not work otherwise. But since we know ahead of time that our fields are untranslatable we can skip this discovery and the corresponding performance penalty.

I am using what seems to be called the "original values" in the Field API. It has very limited API support, since we are supposed to be using the slow `::get()` method. As far as I have seen it is used for the following:

* The public method `ContentEntityBase::updateOriginalValues()` which is called in presave hooks.  This seems to be the only public method, there is no corresponding getter for us to use.
* Construction of a new entity (ref `ContentEntityBase::__construct()` - it has the line `$this->values = $values` which is as fast as you can possibly get :grin: ).
* Initial population of the field values from the database records (`ContentEntityBase::mapFromStorageRecords()` and `SqlContentEntityStorage::mapFromStorageRecords()`).
* Serialization and deserialization (ref. `ContentEntityBase::__sleep()`).

> If there is a performance hit to using get() vs using the raw value, shouldn't this be solved higher up (in core)?

I have been thinking about this and there might be ways, but it would always bypass the dynamic nature of the field definitions. From the core perspective they want a single unified API and you simply _have_ to load the field definitions and incur the performance penalty in order to get the data about the field. Only when you define custom getters in a custom entity class like we do we can avoid this and also get other benefits like strict typing etc. So we could perhaps propose a trait that can be used for custom getters?

> Are we 100% sure these entities will never be translated (and if so shouldn't we prevent it from being marked as translatable - if this isn't the case yet)?

None of them are human readable strings, so it does not make sense to translate them. These are values like the creation date, the user ID, and the only strings are machine names like entity types, bundle IDs and role references.